lib/mutant/isolation.rb

```diff
@@ -1,62 +1,84 @@
 module Mutant
   # Module providing isolation
-  module Isolation
+  class Isolation
     Error = Class.new(RuntimeError)
 
-    module None
-
+    class None < self
       # Call block in isolation
       #
       # @return [Object]
       #
       # @raise [Error]
       #   if block terminates abnormal
-      def self.call(&block)
-        block.call
+      def call
+        yield
       rescue => exception
         raise Error, exception
       end
-    end
+    end # None
+
+    # Isolation via the fork(2) systemcall.
+    #
+    # We do inject so many globals and common patterns to make this unit
+    # specifiable without mocking the globals and more important: Not having
+    # mutations that bypass mocks into a real world side effect.
+    class Fork < self
+      include Anima.new(:process, :stderr, :stdout, :io, :devnull, :marshal)
 
-    module Fork
+      # Prevent mutation from process.fork in #call to call Kernel#fork
+      undef_method :fork
 
       # Call block in isolation
       #
-      # This isolation implements the fork strategy.
-      # Future strategies will probably use a process pool that can
-      # handle multiple mutation kills, in-isolation at once.
-      #
       # @return [Object]
       #   returns block execution result
       #
       # @raise [Error]
       #   if block terminates abnormal
-      #
-      # rubocop:disable MethodLength
-      def self.call(&block)
-        IO.pipe(binmode: true) do |reader, writer|
-          writer.binmode
-          begin
-            pid = Process.fork do
-              File.open(File::NULL, File::WRONLY) do |file|
-                $stderr.reopen(file)
-                reader.close
-                writer.write(Marshal.dump(block.call))
-                writer.close
-              end
-            end
-
-            writer.close
-            Marshal.load(reader.read)
-          ensure
-            Process.waitpid(pid) if pid
-          end
+      def call(&block)
+        io.pipe(binmode: true) do |reader, writer|
+          parent(reader, writer, &block)
         end
       rescue => exception
         raise Error, exception
       end
 
-    end # Fork
+      # Handle parent process
+      #
+      # @param [IO] reader
+      # @param [IO] writer
+      #
+      # @return [undefined]
+      def parent(reader, writer, &block)
+        pid = process.fork do
+          child(reader, writer, &block)
+        end
+
+        writer.close
+        marshal.load(reader)
+      ensure
+        process.waitpid(pid) if pid
+      end
 
+      # Handle child process
+      #
+      # @param [IO] reader
+      # @param [IO] writer
+      #
+      # @return [undefined]
+      def child(reader, writer)
+        reader.close
+        writer.binmode
+
+        devnull.call do |null|
+          stderr.reopen(null)
+          stdout.reopen(null)
+          writer.syswrite(marshal.dump(yield))
```

Comment:
I would recommend changing this so that it makes it more clear the specific action that is causing `stderr` and `stdout` to be be redirected:

```ruby
def child(reader, writer)
  output = devnull.call do |null|
    stderr.reopen(null)
    stdout.reopen(null)
    yield
  end

  reader.close
  writer.binmode.syswrite(marshal.dump(output))
  writer.close
end
```